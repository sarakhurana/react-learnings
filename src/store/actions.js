import axios from "axios";

export const INCREMENT = "INCREMENT";
export const DECREMENT = "DECREMENT";
export const ADD = "ADD";
export const SUBTRACT = "SUBTRACT";
export const CREATE_POST = "CREATE_POST";
export const EDIT_POST = "EDIT_POST";
export const GET_POSTS = "GET_POST";
export const DELETE_POST = "DELETE_POST";
export const GET_POSTS_FAILURE = "GET_POST_FAILURE";
export const FAVOURITE_POST = "FAVOURITE_POST";
export const UNFAVOURITE_POST = "UNFAVOURITE_POST";

export const getPosts = (posts) => ({
  type: GET_POSTS,
  payload: posts,
});

export const getPostFailure = () => ({
  type: GET_POSTS_FAILURE,
});

export const fetchPosts = async (dispatch) => {
  try {
    const response = await axios.get("http://localhost:3000/posts");
    const posts = response.data;
    dispatch(getPosts(posts));
  } catch (e) {
    dispatch(getPostFailure());
  }
};

export const createPost = (post) => ({
  type: CREATE_POST,
  payload: post,
});

export const deletePost = (index) => ({
  type: DELETE_POST,
  payload: index,
});

export const favouritePost = (index) => ({
  type: FAVOURITE_POST,
  payload: index,
});

export const editPost = (post) => ({
  type: EDIT_POST,
  payload: post
})