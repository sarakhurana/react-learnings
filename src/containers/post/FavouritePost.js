import { useContext, useEffect, useRef, useState } from "react";
import { Context } from "../../context/Context";
import { favouritePost } from "../../store/actions";
import classnames from "classnames";
import "./favouritePost.css";

const FavouritePost = ({postId}) => {
  const { dispatch } = useContext(Context);
  const [isFavourite, setFavourite] = useState(false);

  const handleFavouriteClick=()=>{
    setFavourite(!isFavourite)
    dispatch(favouritePost(postId));
  };
  return (
    <div
      className={classnames("btn-favourite", { "btn-favourite-active": isFavourite })}
      onClick={() =>handleFavouriteClick()}
    >
    </div>
  );
};

export default FavouritePost;
