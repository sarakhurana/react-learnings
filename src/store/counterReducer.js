import * as actionTypes from './actions'

const initialState = {
  counter: 0,
};

const counterReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case actionTypes.INCREMENT:
      return { ...state, counter: state.counter + 1 };
    case actionTypes.ADD:
      return { ...state, counter: state.counter + payload };
    case actionTypes.DECREMENT:
      return { ...state, counter: state.counter - 1 };
    case actionTypes.SUBTRACT:
      return { ...state, counter: state.counter - payload };
    default:
      return state;
  }
};

export default counterReducer;
