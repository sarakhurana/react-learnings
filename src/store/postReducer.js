import * as actionTypes from "./actions";

const initialState = {
  posts: [],
};

const postReducer = (state = initialState, action) => {
  const { type, payload } = action;
  switch (type) {
    case actionTypes.CREATE_POST: {    
      state.posts.unshift(payload);
      return state
    }
    case actionTypes.GET_POSTS:
      return { ...state, posts: state.posts.concat(payload) };
    case actionTypes.GET_POSTS_FAILURE:
      return state;
    case actionTypes.DELETE_POST: {
      const updatedArray = state.posts.filter(
        result => result.postId !== payload
      );
      return { ...state, posts: updatedArray };
    }
      case actionTypes.FAVOURITE_POST: {
        const posts=state.posts;
        const index=posts.findIndex(post=>post.postId===payload)
        const newStatus= !posts[index].isFavourite
        const updatedState=state.posts;
        updatedState[index]={...state.posts[index], isFavourite:newStatus}
        return {...state, posts:updatedState};
    }
    case actionTypes.EDIT_POST: {
      const posts=state.posts;
      const index=posts.findIndex(post=>post.postId===payload.postId)
      const updatedState=state.posts;
      updatedState[index]={ ...state.posts[index], title: payload.title, body: payload.body}
      return {...state, posts:updatedState};
  }
    default:
      return state;
  }
};

export default postReducer;
