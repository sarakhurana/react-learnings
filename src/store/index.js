import { combineReducers } from "redux";
import postReducer from "./postReducer";
import counterReducer from "./counterReducer";

const rootReducer = combineReducers({
  counter: counterReducer,
  post: postReducer,
});
export default rootReducer;
