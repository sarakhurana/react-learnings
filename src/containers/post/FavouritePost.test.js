import { render } from "@testing-library/react";
import FavouritePost from "./favouritePost";
import configureMockStore from "redux-mock-store";
import { favouritePost, FAVOURITE_POST } from "../../store/actions";

const mockStore = configureMockStore();

describe("favouritePost", () => {
  let store;
  beforeEach(() => {
    store = mockStore({ post: ["abc"] });
  });

  it("should render", () => {
    const { asFragment } = render(<FavouritePost />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should dispatch favouritePost", () => {
    store.dispatch(favouritePost(1));
    expect(store.getActions()).toEqual([{ type: FAVOURITE_POST, payload: 1 }]);
  });
});
