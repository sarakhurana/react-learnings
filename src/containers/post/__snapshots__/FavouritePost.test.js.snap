// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`favouritePost should render 1`] = `
<DocumentFragment>
  <div
    class="btn-favourite"
  />
</DocumentFragment>
`;
